src/mare/compiler/jumps.cr

```diff
@@ -11,87 +11,232 @@
       node
     end
 
     def touch(node : AST::Identifier)
-      if node.value == "error!"
-        # An identifier is an always error if it is the special case of "error!".
-        @analysis.always_error!(node)
-      elsif node.value[-1] == '!'
+      if node.value[-1] == '!'
         # Otherwise, it is a maybe error if it ends in an exclamation point.
         @analysis.maybe_error!(node)
       end
     end
 
-    def touch(node : AST::Group)
-      if node.terms.any? { |t| @analysis.always_error?(t) }
-        # A group is an always error if any term in it is.
+    def touch(node : AST::Jump)
+      case node.kind
+      when JumpKind::Error
         @analysis.always_error!(node)
-      elsif node.terms.any? { |t| @analysis.maybe_error?(t) }
-        # Otherwise, it is a maybe error if any term in it is.
-        @analysis.maybe_error!(node)
+
+        try_node = @stack.reverse.find(&.is_a?(AST::Try))
+
+        analysis.catch(try_node.not_nil!, node) if try_node
+      when JumpKind::Break, JumpKind::Continue
+        case node.kind
+        when JumpKind::Break
+          @analysis.always_break!(node)
+        when JumpKind::Continue
+          @analysis.always_continue!(node)
+        else
+        end
+
+        loop_node = @stack.reverse.find(&.is_a?(AST::Loop | AST::Yield))
+
+        Error.at node, 
+          "Expected to be used in loops" unless loop_node
+
+        analysis.catch(loop_node.not_nil!, node)
+      when JumpKind::Return
+        @analysis.always_return!(node)
+
+        analysis.catch(@function, node)
+      else
+      end
+    end
+
+    def touch(node : AST::Group)
+      away_flag = 0_u8
+      visitor = AwayVisitor.new(@analysis, away_flag)
+
+      node.terms.each do |t|
+        if @analysis.always_error?(t)
+          @analysis.always_error!(node)
+        elsif @analysis.maybe_error?(t)
+          @analysis.maybe_error!(node)
+        end
+        if @analysis.always_return?(t)
+          @analysis.always_return!(node)
+        elsif @analysis.maybe_return?(t)
+          @analysis.maybe_return!(node)
+        end
+        if @analysis.always_break?(t)
+          @analysis.always_break!(node)
+        elsif @analysis.maybe_break?(t)
+          @analysis.maybe_break!(node)
+        end
+        if @analysis.always_continue?(t)
+          @analysis.always_continue!(node)
+        elsif @analysis.maybe_continue?(t)
+          @analysis.maybe_continue!(node)
+        end
+        begin
+          away_flag |= @analysis.@flags[t][0]
+        rescue
+        end
+
+        @analysis.set_flag(node, away_flag)
+
+        if away_flag > 0 && node.style == ":"
+          visitor.away_flag = away_flag
+          write_unreal_flags = true
+          if flags = @analysis.@flags[t]?
+            write_unreal_flags = false if flags[1]
+          end
+          if write_unreal_flags
+            @analysis.set_flag(t, away_flag, false)
+          end
+          t.accept(@ctx, visitor)
+        end
```

Comment:
I don't really understand what's going on here with real/unreal - what does that mean, and how is it used?

Reply:
Real aways are return error! break and continue. Unreals are all subsequent nodes. It's used in the completeness pass to check if the field write can go away. Also it can be used in the refer pass